### PR TITLE
[fix]: support pytorch SyncBatchNorm under AMP & checkpointing with FSDP

### LINKED ISSUE #617: FSDP and checkpoint_wrapper with support multiple forward passes in sub-modules
## 🐛 Bug

In the context of VISSL, we are training a SWAV model with FSDP and activation checkpointing. The SWAV algorithm requires two forward passes to be performed to compute the loss (each pass deals with images in different resolutions).

When enabling activation checkpointing (using the pattern `FSDP(checkpoint_wrapper(module))`), this leads to one of the following assert during the `run_backward` execution:

**Assert 1**

```
expected to be in states [<TrainingState.BACKWARD_POST: 4>] but current state is TrainingState.IDLE
```

**Assert 2**

```
RuntimeError: expected to be in states [<TrainingState.IDLE: 1>, <TrainingState.BACKWARD_PRE: 3>] but current state is TrainingState.BACKWARD_POST
```

The test cases for both use cases follows.

## Scenario 1 (independant of VISSL)

Use this script to reproduce the **Assert 1**:
- option `-d` enables a double forward pass to compute the loss
- option `-f` enables FSDP in place of DDP to handle the distributed training
- option `-c` enables activation checkpointing `checkpoint_wrapper` 

All combination of options work except `-fdc` pointing to the fact that FSDP and `checkpoint_wrapper` do not work well together when a module trigger 2 forward pass of an underlying sub-module.

```
import argparse
from typing import List, Union

import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.nn as nn
import torch.optim as optim
from fairscale.nn import checkpoint_wrapper
from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP, auto_wrap_bn
from torch.nn.parallel import DistributedDataParallel


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.block1 = nn.Sequential(
            nn.Conv2d(3, 64, kernel_size=3),
            nn.BatchNorm2d(64),
            nn.ReLU(inplace=True),
        )
        self.block2 = nn.Sequential(
            nn.Conv2d(64, 128, kernel_size=3),
            nn.BatchNorm2d(128),
            nn.ReLU(inplace=True),
            nn.AdaptiveAvgPool2d(output_size=(1, 1)),
            nn.Flatten(),
        )
        self.head = nn.Linear(128, 10)

    def forward(self, x: Union[torch.Tensor, List[torch.Tensor]]):
        if isinstance(x, torch.Tensor):
            return self.head(self.block2(self.block1(x)))
        elif isinstance(x, list):
            ys = [self.head(self.block2(self.block1(e))) for e in x]
            return torch.cat(ys, dim=0)


def create_model(with_fsdp: bool, with_checkpoint: bool):
    model = Model()
    if with_fsdp:
        model.block1 = auto_wrap_bn(model.block1, single_rank_pg=False)
        model.block2 = auto_wrap_bn(model.block2, single_rank_pg=False)
        if with_checkpoint:
            model.block2 = checkpoint_wrapper(model.block2)
        model.block1 = FSDP(model.block1)
        model.block2 = FSDP(model.block2)
        model.head = FSDP(model.head)
    else:
        if with_checkpoint:
            model.block2 = checkpoint_wrapper(model.block2)
    return model


def _distributed_worker(gpu_id: int, with_fsdp: bool, double_forward: bool, with_checkpoint: bool):
    torch.cuda.set_device(gpu_id)
    dist.init_process_group(
        backend="nccl", init_method="tcp://127.0.0.1:9099", world_size=2, rank=gpu_id
    )

    torch.manual_seed(0)
    torch.backends.cudnn.deterministic = True
    if double_forward:
        batch = [
            torch.randn(size=(2, 3, 224, 224)).cuda(),
            torch.randn(size=(2, 3, 96, 96)).cuda(),
        ]
    else:
        batch = torch.randn(size=(4, 3, 224, 224)).cuda()

    model = create_model(with_fsdp, with_checkpoint)
    model = model.cuda()

    if with_fsdp:
        model = FSDP(model)
    else:
        model = DistributedDataParallel(model, device_ids=[gpu_id])

    if gpu_id == 0:
        print(model)

    target = torch.LongTensor([0, 1, 2, 3]).cuda()
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.SGD(model.parameters(), lr=0.1, momentum=0.9)

    for iteration in range(3):
        out = model(batch)
        loss = criterion(out, target)
        print("Loss", iteration, ":", loss.item())
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-f", "--fsdp", action="store_const", const=True, default=False)
    parser.add_argument("-d", "--double", action="store_const", const=True, default=False)
    parser.add_argument("-c", "--checkpoint", action="store_const", const=True, default=False)
    args = parser.parse_args()
    mp.spawn(_distributed_worker, (args.fsdp, args.double, args.checkpoint), nprocs=2)
```

## Scenario 2 (VISSL use case)

Use this script to reproduce the **Assert 2**:
- option `-d` enables a double forward pass to compute the loss
- option `-f` enables FSDP in place of DDP to handle the distributed training
- option `-c` enables activation checkpointing `checkpoint_wrapper` 

The main difference with the previous use case is that the model is much bigger (we use a RegNet16). The assert is different but also occurs in `run_backward`.

Again this error occurs only if FSDP is used in conjunction with `checkpoint_wrapper` and if a module triggers 2 forward passes on an underlying sub-module:
- FSDP works fine with `checkpoint_wrapper` in case of single forward pass
- FSDP works fine with double forward passes if not coupled with `checkpoint_wrapper`

```
import argparse

import torch
import torch.distributed as dist
import torch.multiprocessing as mp
import torch.nn as nn
import torch.optim as optim
from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
from hydra.experimental import initialize_config_module, compose
from torch.nn.parallel import DistributedDataParallel

from vissl.models import build_model
from vissl.utils.hydra_config import convert_to_attrdict


def create_config():
    with initialize_config_module(config_module="vissl.config"):
        cfg = compose(
            "defaults",
            overrides=[
                "config=pretrain/swav/swav_8node_resnet",
                "+config/pretrain/swav/models=regnet16Gf",
                "config.SEED_VALUE=2",
                "config.MODEL.AMP_PARAMS.USE_AMP=True",
                "config.MODEL.AMP_PARAMS.AMP_TYPE=pytorch",
                "config.MODEL.SYNC_BN_CONFIG.CONVERT_BN_TO_SYNC_BN=True",
                "config.MODEL.SYNC_BN_CONFIG.SYNC_BN_TYPE=pytorch",
                "config.OPTIMIZER.num_epochs=1",
                "config.OPTIMIZER.param_schedulers.lr.lengths=[0.1, 0.9]",
                "config.OPTIMIZER.use_larc=False",
                "config.LOSS.swav_loss.epsilon=0.03",
                "config.DATA.TRAIN.DATASET_NAMES=[imagenet1k_folder]",
                "config.DATA.TRAIN.DATA_SOURCES=[disk_folder]",
                "config.DATA.TRAIN.BATCHSIZE_PER_REPLICA=16",
                "config.DISTRIBUTED.NCCL_DEBUG=False",
                "config.SLURM.LOG_FOLDER=/checkpoint/qduval/log_runs",
                "config.SLURM.PARTITION=learnfair",
                "config.SLURM.USE_SLURM=true",
                "config.DISTRIBUTED.NUM_NODES=8",
                "hydra.verbose=true",
                "config.MODEL.FSDP_CONFIG.flatten_parameters=True",
                "config.MODEL.FSDP_CONFIG.mixed_precision=False",
                "config.MODEL.FSDP_CONFIG.fp32_reduce_scatter=False",
            ],
        )
    args, config = convert_to_attrdict(cfg)
    return config


def create_model(with_fsdp: bool, with_checkpoint: bool):
    config = create_config()
    if with_checkpoint:
        config["MODEL"]["ACTIVATION_CHECKPOINTING"]["USE_ACTIVATION_CHECKPOINTING"] = True
    if with_fsdp:
        config["MODEL"]["TRUNK"]["NAME"] = "regnet_fsdp"
        config["MODEL"]["HEAD"]["PARAMS"][0][0] = "swav_head_fsdp"
    else:
        config["MODEL"]["TRUNK"]["NAME"] = "regnet_v2"
    return build_model(config["MODEL"], config["OPTIMIZER"])


def create_batch(double_forward: bool):
    torch.manual_seed(0)
    if double_forward:
        return [
            torch.randn(size=(2, 3, 224, 224)).cuda(),
            torch.randn(size=(2, 3, 96, 96)).cuda(),
        ]
    else:
        return torch.randn(size=(4, 3, 224, 224)).cuda()


def _distributed_worker(gpu_id: int, with_fsdp: bool, double_forward: bool, with_checkpoint: bool):
    torch.cuda.set_device(gpu_id)
    dist.init_process_group(
        backend="nccl", init_method="tcp://127.0.0.1:9099", world_size=2, rank=gpu_id
    )

    torch.manual_seed(0)
    torch.backends.cudnn.deterministic = True

    model = create_model(with_fsdp, with_checkpoint)
    model = model.cuda()

    if with_fsdp:
        model = FSDP(model)
    else:
        model = DistributedDataParallel(model, device_ids=[gpu_id])

    if gpu_id == 0:
        print(model)

    criterion = nn.MSELoss()
    optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)

    for iteration in range(3):
        batch = create_batch(double_forward)
        out = sum(o.sum() for o in model(batch)[0])
        loss = criterion(out, torch.tensor(0.).cuda(gpu_id))
        print("Loss", iteration, ":", loss.item())
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("-f", "--fsdp", action="store_const", const=True, default=False)
    parser.add_argument("-d", "--double", action="store_const", const=True, default=False)
    parser.add_argument("-c", "--checkpoint", action="store_const", const=True, default=False)
    args = parser.parse_args()
    mp.spawn(_distributed_worker, (args.fsdp, args.double, args.checkpoint), nprocs=2)
```

## Environment

- Pytorch 1.6 + cu102
- Fairscale 0.3.5
- VISSL master@3a3934b3cfc1802079da85f7ae703497798f85b6
- `module load cuda/11.0 cudnn/v8.0.3.33-cuda.11.0 NCCL/2.8.3-1-cuda.11.0`

## Interested parties

CC: @min-xu-ai @myleott @prigoyal

### PR DESCRIPTION
**Pull request by min-xu-ai**
- there is an issue of sync_bn with checkpoint_wrapper in AMP, particularly due to checkpoint_wrapper sets `track_running_stats` to False, which interacts badly with AMP.
- add force_input_to_fp32 flag to workaround this issue
- this flag can be used to workaround sync batch norm when AMP and checkpoint is used.
- it can also be used in case a layer insist on using fp32 inputs.
- enhanced the unit test so that it covers more types of models as well as runs faster by caching the ddp results.
- also rebalance the tests a bit.

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?
- [x] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)

## What does this PR do?
fixes #617 . (The second test in the issue does not need to be run since we have a double forward & checkpoint test that is even bigger than that test and it doesn't hit any assertions. The bigger test do have memory issues that's separate and is being debugged.)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by min-xu-ai** [comment]
Hey @blefaudeux @QuentinDuval, since this is a bit vision specific, appreciate it if you can take a look.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -201,6 +201,11 @@ class FullyShardedDataParallel(nn.Module):
             GPU OOM during the forward pass. Setting this flag to true will help clearing this
             cache as inner FSDP instances finish part of the forward pass to save GPU memory.
             Default: False
+        force_input_to_fp32 (bool):
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:204]: noob, but what if the inputs are not float to begin with ?

Comment (blefaudeux) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:204]: (I can see that it's actually handled in the conversion code and would not crash, but I'm wondering whether users would be a bit confused)

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1676,18 +1692,25 @@ def _get_default_cuda_device(module: nn.Module) -> torch.device:
     return torch.device("cuda")
 
 
-@torch.no_grad()
-def cast_inputs_to_fp16(*args: Any, **kwargs: Any) -> Tuple[Any, Any]:
+def cast_inputs_to_dtype(to_fp16: bool, no_grad: bool, *args: Any, **kwargs: Any) -> Tuple[Any, Any]:
```
Comment (blefaudeux) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1695]: nit, but it feels a bit confusing, cast_to_dtype is a generic term, but in essence you get a fp16 bool behind that in the interface

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 1 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:204]: Good point! I will update the docstring to say that only float type of tensor are affected.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1695]: Good point. Let me change to "cast_floats_to_right_precision".

**Comment by min-xu-ai** [comment]
@blefaudeux @QuentinDuval, this looks OK to you?

**Review by blefaudeux — APPROVED** [review]
cannot comment too much on FSDP intrinsics but LGTM

**Pull request merged**